portal/portal-service-impl/impl/src/java/org/sakaiproject/portal/service/PortalServiceImpl.java

```diff
@@ -960,10 +960,14 @@ public void addRecentSite(String userId, String siteId) {
 
 		List<String> current = getRecentSites(userId);
 
-		if (current.size() == 3) {
-			// Oldest is last
-			String last = current.toArray(new String[] {})[current.size() - 1];
+		int maxRecentSites = serverConfigurationService.getInt("portal.max.recent.sites", 3);
+		// Clean up excess sites if user has more than the limit
+		while (current.size() >= maxRecentSites && !current.isEmpty()) {
+			// Remove oldest entry (last in the list)
+			String last = current.get(current.size() - 1);
```

Comment:
Could we use `remove` instead of `get` and avoid re-assigning `current`? That should be a bit more efficient and it could avoid a deadlock should the deletion not work.